Use fallback implementation for C++ aligned_alloc for old glibc++

dreamer opened the pull request:
This function was introduced in C++17 and is not available on GCC 5.x or older.  Another version of this function is available since C11, but it's not exposed when including `<cstdlib>`.

This is needed to enable Vulkan renderer for GZDoom builds inside Steam Runtime environment: dreamer/luxtorpeda#35

coelckers commented:
Since this is an external dependency, have you reported it to the original maintainers as well?

alexey-lysiuk commented:
That `aligned_alloc()` function is specific to [glibc](https://linux.die.net/man/3/aligned_alloc). If you are using different standard library implementation (or very old glibc), you are out of luck.
This should be detected properly in [VulkanMemoryAllocator](https://github.com/GPUOpen-LibrariesAndSDKs/VulkanMemoryAllocator). Just checking GCC here version isn’t enough anyway.

dreamer commented:
@alexey-lysiuk I already said in commit message AND in PR description: this part of code does not use [C11 `aligned_alloc`](https://en.cppreference.com/w/c/memory/aligned_alloc), it uses [C++17 `aligned_alloc`](https://en.cppreference.com/w/cpp/memory/c/aligned_alloc) - in fact, it is the only place in GZDoom code that requires C++17 feature. Steam uses glibc of host system for running the game, the only issue is old compiler and old glibc during compilation. I suspect during runtime, only users of very old systems (like Ubuntu 14.04) would be affected.

> Just checking GCC version isn’t enough anyway.

What do you mean? I use this patch [[link]](https://gitlab.com/luxtorpeda/mirrors/gzdoom/commit/75b7da86f6c5a94acabb610660e2c067131e0fc4) to succesfully build GZDoom using GCC 5.4 provided by Steam Runtime build environment [[build log]](https://gitlab.com/luxtorpeda/packages/gzdoom/-/jobs/272307606). Initially I wanted to make it even more specific to my build setup (checking for `STEAM_RUNTIME_SCOUT` define, which I would need to provide to cmake in my build scripts), but I presumed checking compiler version might help other developers as well.

What changes exactly are requested in code review? It's not clear to me.

@coelckers 

> Since this is an external dependency, have you reported it to the original maintainers as well?

I haven't checked if this is an issue for up-to-date VulkanMemoryAllocator yet - I will and update this PR / link to PR to upstream project (if it'll be needed).

On the first glance at code I see this file changed in upstream project and is different than version bundled with GZDoom.

alexey-lysiuk commented:
At the moment, we are using a limited subset of C++14. GZDoom compiles with Vulkan enabled by GCC 4.9 and 5.x just fine. You can check Travis build logs.
You need figure out why your build environment cannot use glibc’s function. Checking compiler version is pretty strange, and it seems to fix compilation occasionally.
Moreover, custom changes to third party libraries should be avoided, as they are dropped usually during update.

dreamer commented:
> At the moment, we are using a limited subset of C++14

From my perspective, you are (completely unintentionally) using a very limited subset of C++17 :)

> GZDoom compiles with Vulkan enabled by GCC 4.9 and 5.x just fine. 

… on Ubuntu 16.04 or newer. I still don't understand why this code compiles when explicitly using C++14 without flag `-faligned-new` (in my opinion, it should not - but maybe I am missing a place where cmake injects this flag).

> You need figure out why your build environment cannot use glibc’s function

To be fair it's not glibc issue - it's glibc++ (if this code was C, the problem would not exist).

> Checking compiler version is pretty strange

I agree; tested several different macros to test feature level support in glibc++ and used the one that I deemed the most appropriate (commit amended).

> Moreover, custom changes to third party libraries should be avoided

I also agree, but you are bundling older version of this library, so I don't know if you plan to update it at some point or not; I made a PR to upstream as well: GPUOpen-LibrariesAndSDKs/VulkanMemoryAllocator#75

alexey-lysiuk commented:
While this version is much better indeed, why did you remove OpenBSD check? Despite being a custom change, it’s already there. I would keep until it will be overwritten occasionally by VulkanMemoryAllocator update :)

dreamer commented:
Argh! I copy-pasted line I changed in upstream project and replaced `#elif` with `#if`, without double checking it afterwards. That's exactly why code reviews are important - thanks for spotting this :)

Commit amended.

kcat commented:
> I already said in commit message AND in PR description: this part of code does not use C11 `aligned_alloc`, it uses C++17 `aligned_alloc`

Rather late on this, but there's no difference, really. Similar to other functions like `malloc` or `log2f`, C++ "adopts" functions and their related headers from C. As noted, C11 first introduced `aligned_alloc` and is thus part of a C11-compliant `stdlib.h` header and libc runtime (like glibc, which glibc++ relies on). C++17 officially adopts this function from C11, the same way C++11 adopted `fesetenv` from C99.

So if your compiler has a C11-compliant `stdlib.h` that declares `aligned_alloc`, and you redeclare/define it because you don't know if it's available, you can end up redeclaring with different attributes or properties which you're not supposed to do (it can mess up what properties or attributes actually get applied). Similarly, If you define it in your program and it's also defined in the C runtime that gets linked in, that breaks the ODR (one definition rule).

My experience with this kind of thing has been, if you need the behavior of a standard function and that function can't be relied upon to exist, make your own version with a different name (or in C++, in your own namespace), which can wrap the standard function if it exists and have a fallback if not. It easily becomes a headache otherwise. Similar to keywords, it's just all around a good idea to not define or declare standard stuff yourself.

dreamer commented:
@kcat unfortunately, in this case, there was a difference. `aligned_alloc` was included in C `stdlib.h` but omitted from C++ `cstdlib` for several years (they basically forgot to add `using ::aligned_alloc;` in there) - glibc++ developers eventually caught that and fixed the issue - that's why I said this problem would not exist if this code was written in C.

This is regular occurence for such functions in C++, but you're likely to see it only if using old build environments and building code using new standards (e.g. in enterprise environments or - int this case - using Steam Runtime). I experienced it e.g. in `math.h`/`cmath` in the past, when C++ started to replace macros with templates but kept old names.

I agree with your position on picking name for fallbacks functions and not redefining standard stuff - it would be nice to have this function namespaced in upstream library (e.g. as `vma::aligned_alloc`), but it is outside of scope of my PR.
